### ISSUE TITLE
cache: an identical prompt is always re-prefilled from scratch (a snapshot at the prompt's own length is refused), and after three repeats even the shared prefix goes cold -- Crow #217's same-prefix retry would pay a full cold prefill (~4 min at 180k tokens)

### ISSUE THREAD

**Issue by nibor1896** [nibor1896/crow-nest#100]
## Symptom

A request whose prompt is **identical** to one the engine just served is re-prefilled from scratch, and after a few repeats even the shared prefix of a longer prompt goes cold. Measured by the #91 replay probe on 2026-09-22 (~19:45 CEST, `decode_out/corruption-replay/baseline.log`), same engine boot, the same three prompts sent 8× each (seeds 0–7):

| point | prompt tokens | cached per seed 0 … 7 |
|---|---|---|
| K=2 | 6,738 | 0 · 0 · 0 · 0 · 0 · 0 · 0 · 0 |
| K=26 | 24,410 | 6,738 · 6,738 · 6,738 · 0 · 0 · 0 · 0 · 0 |
| K=69 | 39,273 | 24,410 · 24,410 · 24,410 · 0 · 0 · 0 · 0 · 0 |

Wall time follows: K=69 seeds 0–2 took 40 / 17 / 20 s, seeds 3–7 45–79 s.

## Evidence

- `engine/src/cache.rs` `reuse_slot`: a snapshot exactly at the new prompt length is refused, test `a_snapshot_at_the_new_length_is_refused` (cache.rs:612) — "`prefill` needs at least one token: it returns the last position's greedy id". A snapshot at `len-1` would be taken (`a_snapshot_one_below_the_new_length_is_taken`), but none is ever made there.
- So an identical re-request can only reuse an OLDER, shorter snapshot (here the previous point's prompt end), and once the slots are recycled by later snapshots (seed ≥ 3 above) it reuses nothing. Which snapshot evicts which is not yet traced (unverified).

## Attribution

crow-nest (cache design). Not a correctness problem — a cost problem that became acute with Crow #217: a degenerate round is now re-requested **on the same prefix** with a fresh seed. At a 180k-token session that is a full cold prefill (~700 tok/s measured → ~4 min) for every retry, which defeats the retry. The same holds for Crow's #151 stream retry and the #210 digest retry-on-same-prefix.

## Research

(From the projects' documentation as recalled when filing -- to be re-verified with sources by whoever implements this.)

- llama.cpp server keeps the last token's logits with the slot's KV state, so an identical prompt re-samples without prefill (`n_past == n_prompt` path re-evaluates at most one token); vLLM's prefix caching also always recomputes the final token of a fully cached prompt to obtain logits (vLLM docs, "Automatic Prefix Caching": the last block is never counted as a full hit).
- For a hybrid model with recurrent (GDN/SSM) state the KV cannot be rewound by one token, so the recompute-last-token trick requires a snapshot taken at `len-1`, or the stored logits.

## Proposed fix

1. **Store the last position's logits (or the sampler-ready row) with each snapshot** taken at a prompt end; `reuse_slot` then accepts a snapshot at exactly the new length and the request goes straight to sampling. Cost: one vocab-sized f32 row per slot (~0.6 MB at 151k vocab).
2. Alternatively / additionally: when a prompt is prefilled cold, **snapshot at `len-1`** before feeding the last token, so the identical re-request prefills one token.
3. Trace and document the slot eviction order that made seeds 3–7 lose even the shared prefix; keep the prompt-end snapshot of the most recent request pinned until a different prefix replaces it.
4. Tests in cache.rs: identical re-request → 0 prefill (or 1 token); the shared prefix survives N identical repeats.

## Expected result

The #91 replay shows `cached == prompt` (or `prompt-1`) for seeds 1–7 at every point; a Crow #217 retry at 180k tokens costs sampling only instead of ~4 minutes of prefill.

**Comment by nibor1896** [comment]
## Research

Re-verified with sources on 2026-09-23 (read, not run), replacing the recalled notes in the ticket body.

- **llama.cpp server** (ggml-org/llama.cpp master e6ab7c1, `tools/server/server-context.cpp`):
  - `[TAG_PROMPT_LOGITS]`: `if (n_past == slot.task->n_tokens() && n_past > 0) { ... "need to evaluate at least 1 token for each active slot" ...; n_past--; }` - an identical prompt re-evaluates the last token to get its logits. The ticket body's claim that llama.cpp stores the last logits with the slot is **wrong**; it recomputes.
  - For hybrid/recurrent memory (our case: GDN) that `n_past--` cannot rewind the state. llama.cpp therefore creates **context checkpoints** of the recurrent state, and splits the tail of the prompt into separate batches so that checkpoints land at `n_tokens - (4 + n_ubatch)` and `n_tokens - 4` ("process the last few tokens of the prompt separately in order to allow for a checkpoint to be created", ref PR #20288). Without a checkpoint it logs "forcing full prompt re-processing due to lack of cache data (likely due to SWA or hybrid/recurrent memory)". An identical prompt on a hybrid model thus re-prefills ~4 tokens.
  - `create_checkpoint`: "replace an existing checkpoint at the same n_tokens instead of appending a duplicate" - a repeat does not fill the checkpoint list with copies of itself. Default `--ctx-checkpoints` 32 per slot (`common/common.h`, `n_ctx_checkpoints = 32`).
  - After picking the restore point: "erase any checkpoints with pos_max > pos_next" - the stale-snapshot defect filed as #101.
- **vLLM** (`vllm/v1/core/kv_cache_manager.py`, main): "When all tokens hit the cache, we must recompute the last token to obtain logits. Thus, set max_cache_hit_length to prompt_length - 1."
- Both recompute because their KV blocks are shared and immutable; neither stores logits. crow-nest holds ONE conversation and snapshots the recurrent state anyway, so the logits row (V = 248,320 f32 = 993,280 B, `geo.rs:22`) can be stored with the snapshot at the same instant.

## Evidence (added): the eviction order, traced

Host-only reproduction on branch t-100 (`cache::tests::the_replay_probe_reuses_every_identical_prompt_and_keeps_the_shared_prefix`, no model, no GPU): the #91 probe's order is point major (`baseline.log`: all 8 seeds of K=2, then K=26, then K=69). Replaying `decide` + the rollback/snapshot bookkeeping of e379361 gives exactly the measured table:

```
left: [[0, 0, 0, 0, 0, 0, 0, 0], [6738, 6738, 6738, 0, 0, 0, 0, 0], [24410, 24410, 24410, 0, 0, 0, 0, 0]]
```

Mechanism: every request snapshots at its prompt end with `rotate_right(1)` (cache.rs:455), even when a slot already holds that position. K=26 seed 0 gives `[24410, 6738, -]`; seeds 1 and 2 cannot use 24410 (`p >= new_len`, cache.rs:193), roll back to 6738, and push two more copies of 24410: `[24410, 24410, 24410]`. The shared prefix 6738 is evicted by duplicates of the refused position; seed 3 finds nothing below 24410 and goes cold, and the cold path's `invalidate` keeps it that way.

## Proposed fix

| option | identical re-request costs | ordinary turn costs | risk |
|---|---|---|---|
| A. store the last logits row + greedy id with each snapshot; accept `S_pos == new_len` for such a slot | rollback + 1 HtoD of 0.99 MB, **0 prefill tokens** | + 1 DtoH of 0.99 MB per snapshot (not measured; `serve.rs` states ~0.3 ms for the same row readback in the #86 path); + 2.98 MB host RAM | the prefill's only output besides the state is `s.logits` row 0 + the greedy id (`gen.rs:3973-3979`); every first-draw path reads exactly that (`arm_sampler` -> `sample_last`, `draw_biased`, #91 logprobs readback `serve.rs:4175`) |
| B. llama.cpp's way: split every prefill at `len-1`, snapshot there, prefill the last token | 1 token prefill | a second `prefill` call on EVERY request; its cost is not measured here, and `gen.rs:3695-3697` notes a small tail chunk is not free | none new, same code path as a warm turn |
| C. only dedupe the snapshots, keep the refusal | still a rollback to the previous prompt end + re-prefill of the rest | none | fixes the "goes cold after 3", not the identical-prompt cost |

**Chosen: A + dedupe + #101.** A is cheaper than llama.cpp's method on both paths and bit-exact (the restored row IS the row the first prefill produced). Together with it:

1. `snapshot` re-uses the slot that already holds `eng.pos` instead of adding a duplicate; otherwise it takes an empty slot, else the one with the SMALLEST position (with #101 all held slots are nested prefixes of `history`, so the smallest is the oldest turn boundary). The new snapshot goes to slot 0, so `SLOT_PROMPT` stays the newest and largest.
2. #101: `rollback` forgets every slot above its point.
3. A slot filled from a slot file (#32 A10) carries no logits row, so it stays under the old rule (`S_pos < new_len`); the slot file format is unchanged.
4. `CROW_VIT_DUMP` requests never take the zero-prefill path (the dump needs the prefill's logits collection).

Out of scope: the per-request duplicate ~9 ms DtoH on an identical re-request (the slot is re-copied rather than skipped; simpler, and it also refreshes a slot restored from file); carrying the logits row in the slot file (#32 format).

## Expected result

- The host replay of the probe gives `[[0, 6738 x7], [6738, 24410 x7], [24410, 39273 x7]]` and ends holding `[39273, 24410, 6738]`.
- Live: the #91 replay shows `cached == prompt` for seeds 1-7 at every point, `prefill 0 of N tok` on the `[cache]` line, and the K=69 seeds 1-7 wall drops from 17-79 s to decode time only.

**Comment by nibor1896** [comment]
## Implementation

Commit **602a29a** on local branch **t-100** (on main e379361), **not pushed, not built into the live serve**. Fixes #100 and #101 together (same slot bookkeeping).

What changed (`engine/src/cache.rs`, `engine/src/bin/serve.rs`):

- `Snapshot` keeps `logits` (`s.logits` row 0, V f32 = 993,280 B) and `greedy: Option<usize>`; `snapshot(eng, prefill_clean, greedy)` copies the row with the state (one more DtoH of 0.99 MB; its wall is **not measured**).
- `reuse_slot_with_logits`: a slot with a row may sit AT the request length. `decide_for(history, ids, allow_exact)`; serve passes `allow_exact = false` only for `CROW_VIT_DUMP` requests.
- serve: `P == prompt.len()` -> `rollback` + `PrefixCache::restore_logits` (row HtoD, returns the stored greedy id) instead of `prefill`. The `[cache]` line then says `WARM ... prefill 0 of N tok`.
- `claim` (dedupe/eviction): the slot already at `eng.pos`, else an empty slot, else the smallest position; moved to slot 0.
- #101: `rollback` forgets every slot above its point.
- A slot restored from a slot file (#32) has no row (`greedy = None`) and keeps the old guard; the slot file format and `Shape::snapshot_bytes` are unchanged.
- Host RAM: +2,979,840 B (3 rows). No VRAM change.
- Docs: `cache.rs` module doc, serve module doc, `docs/architecture.md` (7.4 as-built table, #68 scope note), `docs/long-context-goalmode.md`, `engine/README.md`.

## Verification (host only: no model load, no serve boot, no GPU)

- `the_replay_probe_reuses_every_identical_prompt_and_keeps_the_shared_prefix` replays the #91 probe's order (point major, 3 prompts x 8 seeds, distinct answers) through `decide` + the rollback/snapshot bookkeeping.
  - before (e379361 bookkeeping): `[[0 x8], [6738, 6738, 6738, 0 x5], [24410, 24410, 24410, 0 x5]]` = the measured `baseline.log` table exactly
  - after: `[[0, 6738 x7], [6738, 24410 x7], [24410, 39273 x7]]`, ending with slots `[39273, 24410, 6738]`
- `an_identical_retry_is_served_from_the_prompt_snapshot_and_evicts_nothing`: 10 retries of a 900-token prompt after turns at 300/600: all `cached 900`, slots stay `[900, 600, 300]`.
- `a_rollback_forgets_every_snapshot_above_its_point` (#101): before, `positions()` was `[280, 300, 200]` and the next turn chose the stale 300; after, it chooses 280.
- Plus rule tests for the exact-length rule, dedupe, eviction order, and slot-file slots without a row.
- `cargo test --release` in engine/: **342 passed / 0 failed / 1 ignored** (baseline 335 + 7 new). The ignored test is the GPU-only `attn_98` in vit.rs, not run. `tools/check_env_docs.py` OK (no new env vars).

## Live check pending (robin)

Not run: the GPU was reserved for another agent's pinned-memory work.

1. Start serve from `~/Projects/crow-nest-wt-100/target/release/serve` with the live flags, port 8099.
2. `python3 tools/corruption-replay-probe.py --port 8099 --preset diorama-0922 --session decode_out/sessions/2026-09-22-diorama-rollover/session.json --label t100 --json decode_out/corruption-replay/t100.json`
3. Expected:
   - `prompt N tok (N cached)` for seeds 1-7 at every point; seed 0 of K=26/K=69 `6738` / `24410` cached as before
   - `[cache] WARM ... prefill 0 of N tok` for seeds 1-7
   - K=69 seeds 1-7 wall = decode only (baseline: 17-79 s)
4. Identity check: with greedy (`--sampling '{"temperature":0}'`, `--rounds 2`), round 1 must produce the same ids as round 0 at every point (the restored row is the row round 0's prefill wrote).
5. Risk to watch: the first `decode_step` now runs right after a rollback with no `prefill` in between. Before, a warm request always ran at least one prefill chunk there. The decode graph is dropped by `rollback` and re-captured as usual. That this path produces correct ids is verified by step 4 only.

Open until robin's live replay.

**Comment by nibor1896** [comment]
Live evidence from robin's diorama run 2026-09-23 (serve crow-nest 70a69e3 booted 04:37:47Z, Crow c4f1b3c, 1,220 messages over two rollovers, engine.log 04:37-07:12Z): 6 requests served with `prefill 0 of N tok` (identical prompt re-asked); warm requests reuse the three snapshot slots (e.g. 07:12:19Z: P 149726, snapshots [149726, 148345, 147831], prefill 938 of 150664). The greedy identity check was not run separately.

Closed on robin's request after the live run.
